sn_auditor/src/dag_db.rs

```diff
@@ -382,6 +287,147 @@ impl SpendDagDb {
 
         Ok(())
     }
+
+    pub(crate) async fn dag_crawling_from_genesis(&mut self) {
+        let client = if let Some(client) = &self.client {
+            client.clone()
+        } else {
+            eprintln!("Do not have an attached client");
+            return;
+        };
+
+        let mut addresses;
+        {
+            let genesis_dag = match new_dag_with_genesis_only(&client).await {
+                Ok(dag) => dag,
+                Err(err) => {
+                    eprintln!("Could not create new DAG from genesis: {err}");
+                    return;
+                }
+            };
+
+            addresses = genesis_dag.get_utxos();
+
+            if let Err(err) = self.merge(genesis_dag) {
+                eprintln!("Failed to merge from genesis DAG into our DAG: {err}");
+                return;
+            }
+        }
+
+        let mut utxo_addresses = BTreeMap::new();
+
+        // Use a concurrent approach to process multiple addresses simultaneously
+        loop {
+            if addresses.is_empty() {
+                let re_attempt_address: Vec<_> = utxo_addresses
+                    .iter()
+                    .filter_map(|(address, time_stamp)| {
+                        if *time_stamp > Instant::now() {
+                            None
+                        } else {
+                            Some(*address)
+                        }
+                    })
+                    .collect();
+                addresses.extend(re_attempt_address);
+                utxo_addresses.retain(|_, time_stamp| *time_stamp > Instant::now());
+
+                if addresses.is_empty() {
+                    tokio::time::sleep(REATTEMPT_INTERVAL).await;
+                    continue;
+                }
+            }
+
+            let client_ref = &client;
+            let spends = join_all(addresses.iter().map(|&address| {
+                let client_clone = client_ref.clone();
+                async move { client_clone.crawl_spend_from_network(address).await }
+            }))
+            .await;
+
+            addresses.clear();
+
+            for (address, result) in spends {
+                let spend = match result {
+                    Ok(spend) => Some(spend),
+                    Err(_) => {
```

Comment:
Error management is omitted here: all errors here are treated similarly, this includes `RecordNotFound` where this behavior is indeed wanted, but we might not want this for other errors like `DoubleSpend` or others.